src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,323 @@
+{
+    CELER_EXPECT(r_b > 0);
+    CELER_EXPECT(a > 0);
+    CELER_EXPECT(std::fabs(tmax) < 2 * pi + std::fabs(tmin));
+
+    if (mirror_)
+    {
+        a_ = -a + pi;
+    }
+}
+
+//---------------------------------------------------------------------------//
+
+/*!
+ * Find all roots for involute surfaces that are within the bounds and result
+ * in positive distances. Performed by doing a Regular Falsi Iteration on the
+ * root function, \f[
+ * f(t) = r_b * [v{cos(a+t) + tsin(a+t)} + u{sin(a+t) - tcos(a+t)}] + xv - yu
+ * \f]
+ * where the Regular Falsi Iteration is given by: \f[
+ * tc = [ta*f(tb) - tb*f(ta)] / [f(tb) - f(ta)]
+ * \f]
+ * where \em tc replaces the bound with the same sign (e.g. \em ta and \em tb
+ * ). The initial bounds can be determined by the set of: /f[ {0, beta - a,
+ * beta - a - pi, beta - a + pi, beta - a - 2pi, beta - a + 2pi ...} /f] Where
+ * \em beta is: \f[ beta = arctan(-v/u) \f]
+ */
+CELER_FUNCTION auto
+InvoluteSolver::operator()(Real3 const& pos,
+                           Real3 const& dir) const -> Intersections
+{
+    // Flatten pos and dir in xyz and uv respectively
+    real_type x = pos[0];
+    real_type const y = pos[1];
+    real_type const z = pos[2];
+
+    real_type u = dir[0];
+    real_type const v = dir[1];
+    real_type const w = dir[2];
+
+    if (mirror_)
+    {
+        x = -x;
+        u = -u;
+    }
+
+    // Lambda used for calculating the roots using Regular Falsi Iteration
+    auto root = [](real_type t,
+                   real_type x,
+                   real_type y,
+                   real_type u,
+                   real_type v,
+                   real_type r_b_,
+                   real_type a_) {
+        real_type a = u * std::sin(t + a_) - v * std::cos(t + a_);
+        real_type b = t * (u * std::cos(t + a_) + v * std::sin(t + a_));
+        real_type c = r_b_ * (a - b);
+        return c + x * v - y * u;
+    };
+    /*
+     * Results initalization
+     */
+    Intersections result;
+    // Initial result vector.
+    result = {no_intersection(), no_intersection(), no_intersection()};
+
+    // Return result if particle is travelling along z-axis.
+    if (u == 0 && v == 0)
+    {
+        return result;
+    }
+
+    // Initialize distance vector with root counter j
+    Array<real_type, 3> dist;
+    real_type j = 0;
+
+    // Conversion constant for 2-D distance to 3-D distance
+
+    real_type convert = std::sqrt(ipow<2>(v) + ipow<2>(u) + ipow<2>(w))
+                        / std::sqrt(ipow<2>(v) + ipow<2>(u));
+
+    /*
+     * Define tolerances.
+     * tol_point gives the tolerance level for a point and is set to 1e-7,
+     * account for the floating point error when performing square roots.
+     * tol_conv gives the tolerance for the Regular Falsi iteration,
+     * ensuring that the root found produces a result that within 1e-8 of 0.
+     */
+    real_type const tol_point = 1e-7;
+    real_type const tol_conv = 1e-8;
```

Comment:
I don't think the absolute tolerance should be hardcoded. We Maybe you should use `Tolerance<>::sqrt_quadratic()` (see the `Tolerance` class in `orange/OrangeTypes.hh`) for the relative (convergence) tolerance, and multiply by the radius to get the absolute tolerance. Another approach to the tolerance would be to use the values in `Tolerance<T>::from_default`...